### PR TITLE
Implement logging for MatchBin

### PR DESCRIPTION
**Pull request by rodsan0**
The aim of this PR is to implement a logging framework inside MatchBin that can then be used by other tools, e.g. SignalGP.

### CODE REVIEW

File: source/tools/MatchBin.h [hunk 1]
```diff
@@ -12,6 +12,11 @@
 #ifndef EMP_MATCH_BIN_H
 #define EMP_MATCH_BIN_H
 
+// take care of preprocessored filename
```
Comment (mmore500) [inline source/tools/MatchBin.h:15]: add comment about how to set the default filename with -D

File: source/tools/MatchBin.h [hunk 2]
```diff
@@ -23,15 +28,288 @@
...
+      typename Metric,
+      typename Selector,
+      typename Regulator
+    >
+    friend class emp::MatchBin;
+
+
+    struct LogEntry {
+      query_t query;
+      std::optional<tag_t> maybe_tag;
```
Comment (mmore500) [inline source/tools/MatchBin.h:71]: add comment explaining that it’s maybe_tag for when match request pulls up no results

File: source/tools/MatchBin.h [hunk 3]
```diff
@@ -23,15 +28,288 @@
...
+      LogEntry,
+      size_t,
+      TupleHash<
+        LogEntry
+      >
+    >;
+
+    size_t log_counter;
+    size_t id;
+    logbuffer_t logbuffer;
```
Comment (mmore500) [inline source/tools/MatchBin.h:87]: add comment explaining what this stores

File: source/tools/MatchBin.h [hunk 4]
```diff
@@ -23,15 +28,288 @@
...
+
+    using logbuffer_t = std::unordered_map<
+      LogEntry,
+      size_t,
+      TupleHash<
+        LogEntry
+      >
+    >;
+
+    size_t log_counter;
```
Comment (mmore500) [inline source/tools/MatchBin.h:85]: add comment explaining what this counts

File: source/tools/MatchBin.h [hunk 5]
```diff
@@ -23,15 +28,288 @@
...
+    size_t id;
```
Comment (mmore500) [inline source/tools/MatchBin.h:86]: Add comment explaining this

File: source/tools/MatchBin.h [hunk 6]
```diff
@@ -23,15 +28,288 @@
...
+    logbuffer_t logbuffer;
+
+    #ifdef EMP_LOG_MATCHBIN
+    static constexpr bool logging_enabled = true;
+    #else
+    static constexpr bool logging_enabled = false;
+    #endif
+    bool logging_activated{logging_enabled};
```
Comment (mmore500) [inline source/tools/MatchBin.h:94]: On further reflection, we should probably always initialize this to false and then have the user enable it

File: source/tools/MatchBin.h (hunk 1 above)
Comment (rodsan0) [reply source/tools/MatchBin.h]: Implemented in 93e96aa.

File: source/tools/MatchBin.h (hunk 2 above)
Comment (rodsan0) [reply source/tools/MatchBin.h]: Added in f315207.

File: source/tools/MatchBin.h (hunk 3 above)
Comment (rodsan0) [reply source/tools/MatchBin.h]: Added comment in f91cf2c.

File: source/tools/MatchBin.h (hunk 4 above)
Comment (rodsan0) [reply source/tools/MatchBin.h]: Explained in commit f91cf2c.

File: source/tools/MatchBin.h (hunk 5 above)
Comment (rodsan0) [reply source/tools/MatchBin.h]: Comment added in f91cf2c.
I implemented the variable name change in 0543c72.

File: source/tools/MatchBin.h [hunk 7]
```diff
@@ -23,15 +28,288 @@
...
+
+    emp::ContainerDataFile<logbuffer_t> datafile;
+
+    // setup getter functions
+    std::function<query_t(const std::pair<const LogEntry, size_t>)> get_query_log = [](const std::pair<const LogEntry, size_t> datapoint){ return datapoint.first.query; };
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h]: Made lines shorter in 695471c.

File: source/tools/MatchBin.h [hunk 8]
```diff
@@ -23,15 +28,288 @@
...
+    #ifdef EMP_LOG_MATCHBIN
+    static constexpr bool logging_enabled = true;
+    #else
+    static constexpr bool logging_enabled = false;
+    #endif
+    bool logging_activated{logging_enabled};
+
+    emp::ContainerDataFile<logbuffer_t> datafile;
+
+    // setup getter functions
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h:104]: Added comment in 5a2a6d5.

Sadly, non-static members cannot be `auto`. I did make these functions shorter by using `using`

File: source/tools/MatchBin.h [hunk 9]
```diff
@@ -23,15 +28,288 @@
...
+      if (datapoint.first.maybe_tag) {
+        return emp::to_string(*datapoint.first.maybe_tag);
+      } else {
+        return emp::to_string("");
+      }
+    };
+    std::function<size_t(const std::pair<const LogEntry, size_t>)> get_hit_count_log = [](const std::pair<const LogEntry, size_t> datapoint){ return datapoint.second; };
+    std::function<std::string(const std::pair<const LogEntry, size_t>)> get_logbuffer_type = [](const std::pair<const LogEntry, size_t> datapoint) { return datapoint.first.buffer; };
+
+    size_t MakeID() {
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h:126]: After fbdc0bf, `MakeID()` now has a docstring.

File: source/tools/MatchBin.h [hunk 10]
```diff
@@ -23,15 +28,288 @@
...
+      static std::atomic<int> counter{0};
+      return counter++;
+    }
+
+    void SetupDatafile() {
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h:132]: Also implemented in fbdc0bf.

File: source/tools/MatchBin.h [hunk 11]
```diff
@@ -23,15 +28,288 @@
...
+      datafile.AddContainerFun(get_query_log, "query", "query");
+      datafile.AddContainerFun(get_tag_log, "tag", "tag");
+      datafile.AddContainerFun(get_hit_count_log, "hit_count", "hit_count");
+      datafile.AddContainerFun(get_logbuffer_type, "matchtype", "Type of match");
+
+      datafile.PrintHeaderKeys();
+    }
+
+    // logs only if logging is enabled
+    void LogMatch(const query_t& query, const tag_t& tag, const std::string& buffer) {
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h:146]: Added in fbdc0bf.

File: source/tools/MatchBin.h [hunk 12]
```diff
@@ -23,15 +28,288 @@
...
+      if constexpr (logging_enabled) {
+        if (logging_activated) {
+          LogEntry logentry{query, tag, buffer};
+          ++logbuffer[logentry];
+        }
+      }
+    }
+    void LogMiss(const query_t& query, const std::string& buffer) {
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h:155]: Added in fbdc0bf.

File: source/tools/MatchBin.h [hunk 13]
```diff
@@ -23,15 +28,288 @@
...
+      /// Clear the logs
+      /// This method must be called manually each time.
+      void ClearLogBuffer() {
+        if constexpr (logging_enabled) {
+          logbuffer.clear();
+        }
+      };
+
+      /// Write log buffer to file
+      void WriteLogBuffer() {
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h]: Done 84ef9cf

File: source/tools/MatchBin.h [hunk 14]
```diff
@@ -23,15 +28,288 @@
...
+      /// Write logbuffer to file and then clear it
+      void FlushLogBuffer() {
+        if constexpr (logging_enabled) {
+          WriteLogBuffer();
+          ClearLogBuffer();
+        }
+      }
+
+      /// Enable logging
+      /// Will have no effect if EMP_MATCHBIN_LOG was not set at compile-time
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h:207]: Thanks :3

File: source/tools/MatchBin.h [hunk 15]
```diff
@@ -23,15 +28,288 @@
...
+
+      /// Disable logging
+      /// Will have no effect if EMP_MATCHBIN_LOG was not set at compile-time
+      void Deactivate() { logging_activated = false; }
+
+      /// Set logging to given argument
+      /// Will have no effect if EMP_MATCHBIN_LOG was not set at compile-time
+      void Set(bool log) { logging_activated = log; }
+
+      constexpr bool IsEnabled() const { return logging_enabled; }
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h:220]: Added in d806946.

File: source/tools/MatchBin.h [hunk 16]
```diff
@@ -23,15 +28,288 @@
...
+      bool IsActivated() const { return logging_activated; }
+
+      /// Access the data in logbuffer.
+      logbuffer_t GetLogBuffer() { return logbuffer; }
+
+      // get this instance's ID
+      int GetID() const { return id; }
+
+      template <typename ...ARGS>
+      void EmplaceDataFile(ARGS&&... arguments) {
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h:236]: Done in d806946.

File: source/tools/MatchBin.h [hunk 17]
```diff
@@ -23,15 +28,288 @@
...
+      void ClearRegulated() {
+        if constexpr (cache_available) cache_regulated.clear();
+      }
+
+      /// Reset the Selector cache for raw scores.
+      void ClearRaw() {
+        if constexpr (cache_available) cache_raw.clear();
+      }
+
+      void Clear() {
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h:285]: Sure! 1e6c5ae

File: source/tools/MatchBin.h [hunk 18]
```diff
@@ -615,9 +844,17 @@ namespace emp {
     /// Load MatchBin state
     void SetState(const state_t & state_) {
       state = state_;
-      ClearCache();
+      cache.Clear();
     }
 
+    emp::internal::MatchBinLog<query_t, tag_t>& GetLog() override { return log; }
+
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h:904]: See a27c8dc

File: tests/test_matchbin_logging.cc [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+
+  // can we reenable it?
+  bin.log.Activate();
+  REQUIRE(bin.log.IsActivated());
+
+  // what about setting it to a bool?
+  bin.log.Set(false);
+  REQUIRE(!bin.log.IsActivated());
+  }
+  {
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply tests/test_matchbin_logging.cc:143]: Not anymore! c495084

File: source/tools/MatchBin.h (hunk 6 above)
Comment (rodsan0) [reply source/tools/MatchBin.h]: Done in 41cfb50 and d0b9d53.

**Comment by rodsan0** [comment]
TODO tests:

- test logging a raw and regulated miss
- test returning reference to logging instance (in container of ptrs to MatchBinBase)

- test raw cache size
- test matching with cache disabled
- test clearing cache

File: source/tools/MatchBin.h [hunk 20]
```diff
@@ -219,20 +471,7 @@ namespace emp {
     /// Calling with n = 0 means delegate choice for how many values to return
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h]: done! e29ec08 for implementation and 6daa718 for tests.

File: source/tools/MatchBin.h [hunk 21]
```diff
@@ -23,15 +28,288 @@
...
+        std::unique_lock lock(cache_regulated_mutex);
+        cache_regulated.emplace(query, result);
+      }
+      void CacheRaw(const query_t& query, const cache_state_t& result) {
+        if (CountRaw(query) != 0) return;
+        std::unique_lock lock(cache_raw_mutex);
+        cache_raw.emplace(query, result);
+      }
 
+      // TODO: fix the type name
```
[the remark this answers is not in the document]
Comment (rodsan0) [reply source/tools/MatchBin.h]: 920ddd7 now we did!

**Pull request merged**